spanner/row.go

```diff
@@ -378,3 +386,123 @@ func (r *Row) ToStructLenient(p interface{}) error {
 		true,
 	)
 }
+
+// SelectAll scans rows into a slice (v)
+func SelectAll(rows Iterator, v interface{}, options ...DecodeOptions) error {
+	if rows == nil {
+		return fmt.Errorf("rows is nil")
+	}
+	if v == nil {
+		return fmt.Errorf("p is nil")
+	}
+	vType := reflect.TypeOf(v)
+	if k := vType.Kind(); k != reflect.Ptr {
+		return errToStructArgType(v)
+	}
+	sliceType := vType.Elem()
+	if reflect.Slice != sliceType.Kind() {
+		return errNotASlicePointer()
+	}
+	sliceVal := reflect.Indirect(reflect.ValueOf(v))
+	itemType := sliceType.Elem()
+	s := &decodeSetting{}
+	for _, opt := range options {
+		opt.Apply(s)
+	}
+
+	isPrimitive := itemType.Kind() != reflect.Struct
+	var pointers []interface{}
+	var err error
+	if err := rows.Do(func(row *Row) error {
+		sliceItem := reflect.New(itemType).Elem()
+		if len(pointers) == 0 {
+			if isPrimitive {
+				if len(row.fields) > 1 {
+					return errTooManyColumns()
+				}
+				pointers = []interface{}{sliceItem.Addr().Interface()}
+			} else {
+				if pointers, err = structPointers(sliceItem, row.fields, s.Lenient); err != nil {
+					return err
+				}
+			}
+		}
+		if len(pointers) == 0 {
+			return nil
+		}
+		err := row.Columns(pointers...)
+		if err != nil {
+			return err
+		}
+		if len(pointers) > 0 {
+			dst := sliceItem.Addr().Interface()
+			for i, p := range pointers {
+				reflect.ValueOf(dst).Elem().Field(i).Set(reflect.ValueOf(p).Elem())
+			}
+		}
+		sliceVal.Set(reflect.Append(sliceVal, sliceItem))
```

Comment:
would be great if we could preallocate the whole resultset, but IIRC this is not possible today because the spanner client does not know how many rows are there in the resultset until the end (not sure if this is a fundamental issue, or it can be improved)